### PR TITLE
bugfix: fix data source initialize more times

### PR DESCRIPTION
**Pull request by l81893521**
### Ⅰ. Describe what this PR did
fix data source initialize more than one time cause the data source instance is not a single instance.

### CODE REVIEW

File: core/src/main/java/io/seata/core/store/db/DataSourceGenerator.java [hunk 1]
```diff
@@ -25,10 +25,9 @@
 public interface DataSourceGenerator {
 
     /**
-     * create DataSource from config
-     *
+     * get the data source
      * @return data source
      */
-    DataSource generateDataSource();
+    DataSource get();
```
Comment (jsbxyyx) [inline core/src/main/java/io/seata/core/store/db/DataSourceGenerator.java:31]: retain generateDataSource() ?

Comment (zjinlei) [reply core/src/main/java/io/seata/core/store/db/DataSourceGenerator.java:31]: +1

Comment (l81893521) [reply core/src/main/java/io/seata/core/store/db/DataSourceGenerator.java:31]: but generate() already in the AbstractDataSourceGenerator.generate(), this method just only get the data source.

File: core/src/main/java/io/seata/core/store/db/AbstractDataSourceGenerator.java [hunk 2]
```diff
@@ -60,6 +65,24 @@
 
     private static final long DEFAULT_DB_MAX_WAIT = 5000;
 
+    /**
+     * generate the data source
+     */
+    public AbstractDataSourceGenerator() {
+        this.dataSource = this.generate();
```
Comment (zjinlei) [inline core/src/main/java/io/seata/core/store/db/AbstractDataSourceGenerator.java:72]: recommended to implement io.seata.common.executor.Initialize

Comment (l81893521) [reply core/src/main/java/io/seata/core/store/db/AbstractDataSourceGenerator.java:72]: ok

**Comment by slievrly** [comment]
@l81893521 Whether to need to adjust the connection pool of the default number of connections?

**Comment by l81893521** [comment]
> @l81893521 Whether to need to adjust the connection pool of the default number of connections?

ok, DEFAULT_DB_MAX_CONN already change to 20

File: server/src/main/java/io/seata/server/store/DbcpDataSourceProvider.java [hunk 3]
```diff
@@ -49,6 +49,11 @@ public DataSource generateDataSource() {
         ds.setTestWhileIdle(true);
         ds.setValidationQuery(getValidationQuery(getDBType()));
         ds.setConnectionProperties("useUnicode=yes;characterEncoding=utf8;socketTimeout=5000;connectTimeout=500");
-        return ds;
+        setDataSource(ds);
+    }
+
+    @Override
+    public DataSource provide() {
+        return getDataSource();
```
Comment (zjinlei) [inline server/src/main/java/io/seata/server/store/DbcpDataSourceProvider.java:57]: common method, move to abstract class.

File: server/src/test/java/io/seata/server/store/db/DbcpDataSourceProviderTest.java [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+
+/**
+ * @author will
+ */
+public class DbcpDataSourceProviderTest {
+
+    private final String datasourceType = "dbcp";
+
+    @Test
+    public void testDbcpDataSourceProvider() {
```
Comment (slievrly) [inline server/src/test/java/io/seata/server/store/db/DbcpDataSourceProviderTest.java:33]: You can put the test case in the abstract class

File: server/src/main/java/io/seata/server/store/DbcpDataSourceProvider.java (hunk 3 above)
Comment (l81893521) [reply server/src/main/java/io/seata/server/store/DbcpDataSourceProvider.java]: fixed.

File: server/src/test/java/io/seata/server/store/db/DbcpDataSourceProviderTest.java (hunk 4 above)
Comment (l81893521) [reply server/src/test/java/io/seata/server/store/db/DbcpDataSourceProviderTest.java]: fixed.

**Comment by ph3636** [comment]
Pay attention to the package order

File: server/src/test/java/io/seata/server/store/db/AbstractDataSourceProviderTest.java [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+import org.junit.jupiter.api.Test;
+
+import javax.sql.DataSource;
+
+/**
+ * @author: will
+ */
+public class AbstractDataSourceProviderTest {
+
+    private final String dbcpDatasourceType = "dbcp";
```
Comment (jsbxyyx) [inline server/src/test/java/io/seata/server/store/db/AbstractDataSourceProviderTest.java:30]: merge dataSourceType to array. for each test dataSource not null?

**Pull request merged**